docs/src/design/stack/main.md

```diff
@@ -0,0 +1,226 @@
+
+### Stack depth constraints
+To make sure stack depth column $b_0$ is updated correctly, we need to impose the following constraints:
+
+| Condition   | Constraint       | Description |
+| ----------- | ---------------- | ----------- |
+| $f_{shr}=1$ | $b'_0 = b_0 + 1$ | When the stack is shifted to the right, stack depth should be incremented by $1$. |
+| $f_{shl}=1$ <br> $f_{ov}=1$ | $b'_0 = b_0 - 1$ | When the stack is shifted to the left and the overflow table is not empty, stack depth should be decremented by $1$. |
+| otherwise   | $b'_0 = b_0$     | In all other cases, stack depth should not change. |
+
+We can combine the above constraints into a single expression as follows:
+
+>$$
+b'_0 - b_0 + f_{shl} \cdot f_{ov} - f_{shr} = 0 \text{ | degree } = 7
+$$
+
+### Overflow table constraints
+
+When the stack is shifted to the right, a tuple $(k_0, s_{15}, b_1)$ should be added to the overflow table. We will denote value of the row to be added to the table as follows:
+
+$$
+v = \alpha_0 + \alpha_1 \cdot k_0 + \alpha_2 \cdot s_{15} + \alpha_3 \cdot b_1
+$$
+
+When the stack is shifted to the left, a tuple $(b_1, s'_{15}, b'_1)$ should be removed from the overflow table. We will denote value of the row to be removed from the table as follows.
+
+$$
+u = \alpha_0 + \alpha_1 \cdot b_1 + \alpha_2 \cdot s'_{15} + \alpha_3 \cdot b'_1
+$$
+
+Using the above variables, we can ensure that right and left shifts update the overflow table correctly by enforcing the following constraint:
+
+>$$
+p_1' \cdot (u \cdot f_{shl} \cdot f_{ov} + 1 - f_{shl} \cdot f_{ov}) = p_1 \cdot (v \cdot f_{shr} + 1 - f_{shr}) \text{ | degree } = 9
+$$
+
+We need to add $1$ and subtract the relevant flags from each side to ensure that when none of the flags is set to 1, the above constraint reduces to $p_1' = p_1$.
+
+Notice that in the case of the left shift, the constraint forces the prover to set the next values of $s_{15}$ and $b_1$ to values $t_1$ and $t_2$ of the row removed from the overflow table.
+
+In case of a right shift, we also need to make sure that the next value of $b_1$ is set to the current value of $k_0$. This can be done with the following constraint:
+
+>$$
+f_{shr} \cdot (b'_1 - k_0) = 0 \text{ | degree } = 8
+$$
+
+In case of a left shift, when the overflow table is empty, we need to make sure that a $0$ is "shifted in" from the right (i.e., $s_{15}$ is set to $0$). This can be done with the following constraint:
+
+>$$
+f_{shl} \cdot (1 - f_{ov}) \cdot s_{15}' = 0 \text{ | degree } = 8
+$$
+
+### Stack top constraints
+We also need to make sure the the values in the columns representing the top $16$ items of the stack are updated correctly. In general, these constraints will be operation-specific (i.e., different operation would affect the top of the stack differently). But we can illustrate the general approach on the example of two operations: `DUP` and `DROP`.
```

Comment:
nit: This should be "We also need to make sure that the values...". Also "...different operations would affect..."